src/__tests__/hooks/useOnClickOutside.test.js

```diff
@@ -0,0 +1,66 @@
+import { vi } from 'vitest';
+import { act, renderHook } from '@testing-library/react-hooks';
+import useOnClickOutside from '../../hooks/useOnClickOutside';
+
+describe('useOnClickOutside', () => {
+  it('should call the handler when clicking outside the ref', () => {
+    const ref = { current: document.createElement('div') };
+    const handler = vi.fn();
+
+    renderHook(() => useOnClickOutside(ref, handler));
+
+    act(() => {
+      const outsideElement = document.createElement('div');
+      document.body.appendChild(outsideElement);
+
+      const event = new MouseEvent('mousedown', {
+        bubbles: true,
+        cancelable: true,
+      });
+      outsideElement.dispatchEvent(event);
+    });
+
+    expect(handler).toHaveBeenCalled();
+  });
+
+  it('should not call the handler when clicking inside the ref', () => {
+    const ref = { current: document.createElement('div') };
+    const handler = vi.fn();
+
+    renderHook(() => useOnClickOutside(ref, handler));
+
+    act(() => {
+      const insideElement = document.createElement('div');
+      ref.current.appendChild(insideElement);
+
+      const event = new MouseEvent('mousedown', {
+        bubbles: true,
+        cancelable: true,
+      });
+      insideElement.dispatchEvent(event);
+    });
+
+    expect(handler).not.toHaveBeenCalled();
+  });
+
+  it('should not call the handler after hook is unmounted', () => {
+    const ref = { current: document.createElement('div') };
+    const handler = vi.fn();
+
+    const { unmount } = renderHook(() => useOnClickOutside(ref, handler));
+
+    act(() => {
+      const insideElement = document.createElement('div');
+      ref.current.appendChild(insideElement);
```

Comment:
I think we should use the `outsideElement` code here instead as the handler would only be called if the click was on an element outside the one hosting the hook.

Reply:
Pushed through new changes, but I'm not 100% sure that my test is doing the right thing.

I changed this so that the event is dispatched to outerElement after unmounting the ref element hosting the hook, but for some reason it only works if I don't use `act()`.  Keeping the both `unmount()` and `outsideElement.dispatchEvent(event)` inside the `act()` scope caused the test to fail, removing act results in a failed test with the same assertion.

I read the [docs](https://testing-library.com/docs/preact-testing-library/api/#act) and [this](https://github.com/threepointone/react-act-examples/blob/master/sync.md) related post, but I'm still not 100% sure what is going on.

Am I actually testing the correct behavior this way?

@robert-w-gries

Reply:
AB - use RTL debug to check that all DOM elements are present. Can also get more specific with assertions to double check that correct behavior is being tested